### DISCUSSION TITLE
Unable to git push project created in Codespaces

### DISCUSSION THREAD

**Discussion by valhuber** [community/community#27769] [Codespaces] [answered]
I have a project working well in codespaces.  It runs, and I am able to push changes to my repo, so my git credentials appear to be working for that project.

But if I try to create a new project, I am unable to `git push origin master` it - getting 
```
      remote: Permission to PoseyDev/MyProject.git denied to PoseyDev.
      fatal: unable to access 'https://github.com/PoseyDev/MyProject.git/': The requested URL returned error: 403
```

I have documented the exact commands and responses [here](https://valhuber.github.io/ApiLogicServer/Tech-CodeSpaces/#create-from-codespaces-currently-not-working).  I am a naive git user, so this could well be a cockpit error.

**Comment by jfullstackdev** [comment]
is this your another account ? 

![image](https://user-images.githubusercontent.com/47092464/182011408-ad5e6a23-1617-48e4-889c-ba395249f5f1.png)

that might be a confusion because you are using two accounts

**Comment by valhuber** [comment]
It is the account for both.  The first project was a fork, if that matters…

On Sat, Jul 30, 2022 at 10:21 PM jdevstatic ***@***.***>
wrote:

> is this your another account ?
>
> [image: image]
> <https://user-images.githubusercontent.com/47092464/182011408-ad5e6a23-1617-48e4-889c-ba395249f5f1.png>
>
> that might be a confusion because you are using two accounts
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/orgs/community/discussions/27769#discussioncomment-3289225>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ABKSG36J5BKPRP6J5MRK7X3VWYEONANCNFSM55EIZRDQ>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 
Thanks,
Val

510-703-5909  valhuber on skype

**Reply by jfullstackdev** [reply]
> It is the account for both.

what do you mean by that ? I tried testing it in the past, using two accounts, the problem is that, if it's on the same machine, all of the commits will be mixed by the other account, which is actually a mess,

**Comment by valhuber** [comment]
I have my wife’s account PoseyDev for testing “not me”.  It’s the only account I was using for exercise.  I am doing it from her machine, so there is no mixing of accounts as far as I know.

**Reply by jfullstackdev** [reply]
do you use the same computer ? that's what happened to me, even if just once

**Comment by valhuber** [comment]
Hi...

Appreciate your time here... I have re-run my test, eliminating the "other" github account.  Now it's all under my account (valhuber).  The problem still occurs as [documented here](https://valhuber.github.io/ApiLogicServer/Tech-CodeSpaces/).

**Comment by jshorty** [comment, marked as answer]
👋 Hello from the Codespaces team! This is expected behavior at the moment, so not a cockpit error 😄. By default the token your codespace uses to authenticate requests to GitHub is scoped only to repository that it's created with— and so will be forbidden to write to the remote you've set up for your newer repository from within the codespace.

Here is the recommended path for expanding the permissions of a codespace to include other repositories:
https://docs.github.com/en/codespaces/managing-your-codespaces/managing-repository-access-for-your-codespaces

This flips the steps you've documented around, however— it would require you to set up the remote GitHub repo before spinning up the other codespace. We'd be curious to hear if this is a source of friction for you.

You may also find this blog post relevant! https://github.blog/2022-04-20-codespaces-multi-repository-monorepo-scenarios/

**Reply by tyeth** [reply]
Source of friction. If I clone my fork into an existing codespace, and the credentials match, then let me do what I need to do. Alternatively provide a way of adding authorisation for a repo after the codespace is created. I don't want to amend the codespace to access everything in my organisation as that feels like the wrong way per permissions of least responsibility.

**Reply by anaisbetts** [reply]
This decision basically makes Codespaces effectively useless for contributing to Open Source:

1. Create Codespace of upstream repo
2. Make PR
3. Create fork of repo so that you can publish changes
4. You cannot push your changes anywhere - the upstream repo is read-only, and your token has no access to the fork. You now cannot push _anywhere_ and your changes are stuck on the Codespace. I had to move them by copy-pasting diffs.

**Reply by ozyman42** [reply]
this seems incompatible with git submodules.

**Reply by ozyman42** [reply]
Here's how to solve it in case recreating the codespace is not an option 
https://docs.github.com/en/codespaces/troubleshooting/troubleshooting-authentication-to-a-repository#authenticating-to-repositories-that-you-didnt-create-the-codespace-from

**Reply by sarath49** [reply]
Step 5 does not work, so does that mean to follow the instructions smoothly, need me to be Microsoft employee?
https://github.com/skills/code-with-codespaces/blob/main/.github/steps/1-first-codespace.md

**Reply by sarath49** [reply]
Now it works, it looks like we have to wait for few minutes before the push start working, earlier it was all about "remote: Invalid username or password."

**Comment by valhuber** [comment]
Good information, and will surely ponder.   Glad it was not a cockpit error!

**Comment by valhuber** [comment]
I checked her computer, I don’t see my git account.  But this is in codespaces, so that’s a different world, no?

But you made me wonder… did this

***@***.***:/workspaces/Tutorial-ApiLogicProject$ git config --list
credential.helper=/.codespaces/bin/gitcredential_github.sh
user.name=PoseyDev
***@***.***
gpg.program=/.codespaces/bin/gh-gpgsign
core.repositoryformatversion=0
core.filemode=true
core.bare=false
core.logallrefupdates=true
remote.upstream.url=https://github.com/valhuber/Tutorial-ApiLogicProject
remote.upstream.fetch=+refs/heads/*:refs/remotes/upstream/*
branch.main.remote=origin
branch.main.merge=refs/heads/main
remote.origin.url=https://github.com/PoseyDev/Tutorial-ApiLogicProject.git
remote.origin.fetch=+refs/heads/*:refs/remotes/origin/*
***@***.***:/workspaces/Tutorial-ApiLogicProject$



Remember, this all started from forking the project in red, which is in fact my git account.  Do you think that might be the culprit?

Thanks,
Val

**Comment by Yelinz** [comment]
Adding an answer, into this old post due to me being sent to this post through googling this problem.

Nowadays you can "save" your codespace/changes through the UI.
1. Create a fork of the original repo
2. In the fork, click the "Code" button
3. Go to the codespaces tab
4. The "bad" codespace is under "Other Branches"
5. Click export changes to a branch
6. GitHub will move the "bad" codespace into your fork and you can make a pull request

<img width="927" height="728" alt="image" src="https://github.com/user-attachments/assets/2606b1ff-d40b-41f6-815a-eb2ec27c7e46" />
